File: Sources/Scratchpad/Persistence/FileDrawingRepository.swift
import Foundation

/// Persists each display's drawing as its own JSON file, named by that
/// display's stable identifier, under `PersistenceLocations.drawingsDirectory`.
final class FileDrawingRepository: DrawingRepository {
    private let fileManager: FileManager
    private let directory: URL
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(
        fileManager: FileManager = .default,
        directory: URL = PersistenceLocations.drawingsDirectory
    ) {
        self.fileManager = fileManager
        self.directory = directory
    }

    func load(for display: DisplayIdentifier) -> DrawingDocument {
        let url = fileURL(for: display)
        guard let data = try? Data(contentsOf: url) else { return .empty }
        return (try? decoder.decode(DrawingDocument.self, from: data)) ?? .empty
    }

    func save(_ document: DrawingDocument, for display: DisplayIdentifier) throws {
        try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        let data = try encoder.encode(document)
        try data.write(to: fileURL(for: display), options: .atomic)
    }

    func clear(for display: DisplayIdentifier) throws {
        let url = fileURL(for: display)
        guard fileManager.fileExists(atPath: url.path) else { return }
        try fileManager.removeItem(at: url)
    }

    private func fileURL(for display: DisplayIdentifier) -> URL {
        directory.appendingPathComponent("\(display.filenameComponent).json")
    }
}

File: Windows/Scratchpad/Persistence/FileDrawingRepository.cs
using System.IO;
using System.Text.Json;
using Scratchpad.Drawing;
using Scratchpad.Overlay;

namespace Scratchpad.Persistence;

/// Persists each display's drawing as its own JSON file, named by that
/// display's stable identifier, under `PersistenceLocations.DrawingsDirectory`.
/// Mirrors the macOS app's `FileDrawingRepository`.
public sealed class FileDrawingRepository : IDrawingRepository
{
    private readonly string _directory;

    public FileDrawingRepository(string? directory = null)
    {
        _directory = directory ?? PersistenceLocations.DrawingsDirectory;
    }

    public DrawingDocument Load(DisplayIdentifier display)
    {
        var path = FilePath(display);
        if (!File.Exists(path)) return DrawingDocument.Empty;
        try
        {
            var json = File.ReadAllText(path);
            return JsonSerializer.Deserialize<DrawingDocument>(json) ?? DrawingDocument.Empty;
        }
        catch (Exception ex) when (ex is IOException or JsonException or UnauthorizedAccessException)
        {
            return DrawingDocument.Empty;
        }
    }

    public void Save(DrawingDocument document, DisplayIdentifier display)
    {
        Directory.CreateDirectory(_directory);
        var json = JsonSerializer.Serialize(document);
        var path = FilePath(display);
        var tempPath = path + ".tmp";
        File.WriteAllText(tempPath, json);
        File.Move(tempPath, path, overwrite: true);
    }

    public void Clear(DisplayIdentifier display)
    {
        var path = FilePath(display);
        if (File.Exists(path))
        {
            File.Delete(path);
        }
    }

    private string FilePath(DisplayIdentifier display) =>
        Path.Combine(_directory, $"{display.FilenameComponent}.json");
}
